cardinal/config.go

```diff
@@ -43,46 +51,67 @@ var (
-	CardinalLogLevel string `config:"CARDINAL_LOG_LEVEL"`
+	CardinalLogLevel string `config:"CARDINAL_LOG_LEVEL" mapstructure:"CARDINAL_LOG_LEVEL"`
 
 	// CardinalLogPretty Pretty logging, disable by default due to performance impact.
-	CardinalLogPretty bool `config:"CARDINAL_LOG_PRETTY"`
+	CardinalLogPretty bool `config:"CARDINAL_LOG_PRETTY" mapstructure:"CARDINAL_LOG_PRETTY"`
 
 	// RedisAddress The address of the redis server, supports unix sockets.
-	RedisAddress string `config:"REDIS_ADDRESS"`
+	RedisAddress string `config:"REDIS_ADDRESS" mapstructure:"REDIS_ADDRESS"`
 
 	// RedisPassword The password for the redis server. Make sure to use a password in production.
-	RedisPassword string `config:"REDIS_PASSWORD"`
+	RedisPassword string `config:"REDIS_PASSWORD" mapstructure:"REDIS_PASSWORD"`
 
 	// BaseShardSequencerAddress This is the address that Cardinal will use to sequence and recover to/from base shard.
-	BaseShardSequencerAddress string `config:"BASE_SHARD_SEQUENCER_ADDRESS"`
+	BaseShardSequencerAddress string `config:"BASE_SHARD_SEQUENCER_ADDRESS" mapstructure:"BASE_SHARD_SEQUENCER_ADDRESS"`
 
 	// BaseShardRouterKey is a token used to secure communications between the game shard and the base shard.
-	BaseShardRouterKey string `config:"BASE_SHARD_ROUTER_KEY"`
+	BaseShardRouterKey string `config:"BASE_SHARD_ROUTER_KEY" mapstructure:"BASE_SHARD_ROUTER_KEY"`
 
 	// TelemetryEnabled When true, Cardinal will send telemetry to a telemetry agent.
-	TelemetryEnabled bool `config:"TELEMETRY_ENABLED"`
+	TelemetryEnabled bool `config:"TELEMETRY_ENABLED" mapstructure:"TELEMETRY_ENABLED"`
 
 	// TelemetryStatsdAddress The address of a statsd metric agent that will collect stats from Cardinal.
-	TelemetryStatsdAddress string `config:"TELEMETRY_STATSD_ADDRESS"`
+	TelemetryStatsdAddress string `config:"TELEMETRY_STATSD_ADDRESS" mapstructure:"TELEMETRY_STATSD_ADDRESS"`
 
 	// TelemetryTraceAddress The address of an agent that supports the collection of traces (e.g. a DataDog agent).
-	TelemetryTraceAddress string `config:"TELEMETRY_TRACE_ADDRESS"`
+	TelemetryTraceAddress string `config:"TELEMETRY_TRACE_ADDRESS" mapstructure:"TELEMETRY_TRACE_ADDRESS"`
 }
 
 func loadWorldConfig() (*WorldConfig, error) {
+	// Set default config
 	cfg := defaultConfig
 
+	// Setup Viper for world toml config file
+	setupViper()
+
+	// Read the config file
+	if err := viper.ReadInConfig(); err != nil {
+		log.Warn().Err(err).Msg("failed to read config file")
+		log.Info().Msg("using default config values")
+	} else {
+		// Unmarshal the [cardinal] section into the WorldConfig struct
+		if err := viper.Sub("cardinal").Unmarshal(&cfg); err != nil {
+			log.Warn().Err(err).Msg("failed to unmarshal config file")
+			log.Info().Msg("using default config values")
+		}
+	}
+
+	// Override config values with environment variables
+	// This is done after reading the config file to allow for environment variable overrides
 	if err := config.FromEnv().To(&cfg); err != nil {
```

Comment:
We should be able to drop JeremyLoy/config altogether since viper can also read from environment variable no?

Reply:
updated, replaced by viper